lib/recursive-methods.rb

```diff
@@ -1,49 +1,88 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    raise ArgumentError if n <= -1
+    return 1 if n == 0
+    return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+
+    return "" if s.length == 0
+    return s if s.length == 1
+
+    return s[s.length-1].chr + reverse(s[0, s.length-1])
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: 0(n)
 def reverse_inplace(s)
```

Comment:
This works, but it is not in place and essentially the same as the previous method. 

Think about if you adjusted the method signature as follows

```suggestion
# Time complexity: O(n)
# Space complexity: 0(n)
def reverse_inplace(s, low = 0, high = s.length - 1)
```